openshift/oc-mirror
# CLID-10: IDMS/ITMS

# Description

This PR contains implementation for [CLID-10](https://issues.redhat.com//browse/CLID-10) (see jira for detailed design) :
* IDMS and ITMS generation
* Excluded from PR: CatalogSource for operator catalogs will be in a different PR #784

Fixes # [CLID-10](https://issues.redhat.com//browse/CLID-10).

A few precisions regarding IDMS/ITMS generation: [CLID-10](https://issues.redhat.com//browse/CLID-10) (see comments for details):
* By default, ImageDigestMirrorSet is generated, and contains mirrors for images referenced by digest in the release, operator catalog or additional images, helm charts
* By default, ImageTagMirrorSet is generated if at least 1 image referenced by tag was found. ITMS  contains mirrors for images referenced by tag in the release, operator catalog or additional images, helm charts 
* IDMS and ITMS use source and mirror paths scoped to namespace , unless `--max-nested-paths` has been used. In the latter case, repository scope is used.
* V2 doesn't do tag to digest resolution prior to mirroring. In V1, oc-mirror was replacing tags by digests before handing mappings.txt to the oc code, which in turn, created a fake tag (derived from the digest value).  This V1 behavior  is not considered fit for V2 because: 
  * we have ITMS, 
  * we don't call oc, 
  * we don't rebuild catalogs *so far* (the reference for the image is not modified in the declarative config). 
* IDMS/ITMS resource name and filename use a different naming convention than v1: 
  * Filename: `"idms-oc-mirror.yaml"` and `"itms-oc-mirror.yaml"` respectively .
  * resource names:
     * `idms-generic-0`, `itms-generic-0` respectively for IDMS of all additional images by digest and ITMS of all additional images by tag
     * `idms-operator-0`, `itms-operator-0`respectively for IDMS of all operator images by digest and ITMS of all operator images by tag
     * `idms-release-0`, `itms-release-0` respectively for IDMS of all release images by digest and ITMS of all release images by tag (if any!)
 
## Type of change

Please delete options that are not relevant.


# How Has This Been Tested?
Unit tests cover the various cases.
Manual tests:
Prior to testing, create the file clid10.yaml with content:
```yaml
kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v1alpha2
archiveSize: 1
mirror:
  additionalImages:
  - name: "gcr.io/kubebuilder/kube-rbac-proxy:v0.5.0"
  - name: "quay.io/cockroachdb/cockroach-helm-operator:6.0.0"
  - name: "quay.io/helmoperators/cockroachdb:v5.0.3"
  - name: "quay.io/helmoperators/cockroachdb:v5.0.4"
  - name: "quay.io/openshift-community-operators/cockroachdb@sha256:a5d4f4467250074216eb1ba1c36e06a3ab797d81c431427fc2aca97ecaf4e9d8"
  - name: "quay.io/openshift-community-operators/cockroachdb@sha256:d3016b1507515fc7712f9c47fd9082baf9ccb070aaab58ed0ef6e5abdedde8ba"
  - name: "quay.io/openshift-community-operators/cockroachdb@sha256:f42337e7b85a46d83c94694638e2312e10ca16a03542399a65ba783c94a32b63"
```
 Next, perform a mirrorToDisk
```bash
./bin/oc-mirror --v2 -c clid10.yaml file:///home/skhoury/clid-10
```
| Test| Description|Outputs|
|---|---|---|
|`./bin/oc-mirror --v2 -c clid10.yaml --from file:///home/skhoury/clid-10/ docker://localhost:5000/clid10/noflag --dest-tls-verify=false`| Nominal use case - IDMS + ITMS generated - mirrors are namespace scope | idms-nomax.yaml + itms-nomax.yaml|
|`./bin/oc-mirror --v2 -c clid10.yaml --max-nested-paths=2 --from file:///home/skhoury/clid-10/ docker://localhost:5000/clid10/max --dest-tls-verify=false`|generates IDMS + ITMS, with repository scope| idms-max.yaml + itms-max.yaml|

## Expected Outcome
idms-nomax.yaml
```yaml
---
apiVersion: config.openshift.io/v1
kind: ImageDigestMirrorSet
metadata:
  creationTimestamp: null
  name: idms-generic-0
spec:
  imageDigestMirrors:
  - mirrors:
    - localhost:5000/clid10/noflag/openshift-community-operators
    source: quay.io/openshift-community-operators
status: {}
```

idms-max.yaml
```yaml
---
apiVersion: config.openshift.io/v1
kind: ImageDigestMirrorSet
metadata:
  creationTimestamp: null
  name: idms-generic-0
spec:
  imageDigestMirrors:
  - mirrors:
    - localhost:5000/clid10/noflag-openshift-community-operators-cockroachdb
    source: quay.io/openshift-community-operators/cockroachdb
status: {}
```

itms-nomax.yaml
```yaml
---
apiVersion: config.openshift.io/v1
kind: ImageTagMirrorSet
metadata:
  creationTimestamp: null
  name: itms-generic-0
spec:
  imageTagMirrors:
  - mirrors:
    - localhost:5000/clid10/noflag/kubebuilder
    source: gcr.io/kubebuilder
  - mirrors:
    - localhost:5000/clid10/noflag/cockroachdb
    source: quay.io/cockroachdb
  - mirrors:
    - localhost:5000/clid10/noflag/helmoperators
    source: quay.io/helmoperators
status: {}
```

itms-max.yaml
```yaml
---
apiVersion: config.openshift.io/v1
kind: ImageTagMirrorSet
metadata:
  creationTimestamp: null
  name: itms-generic-0
spec:
  imageTagMirrors:
  - mirrors:
    - localhost:5000/clid10/noflag-kubebuilder-kube-rbac-proxy
    source: gcr.io/kubebuilder/kube-rbac-proxy
  - mirrors:
    - localhost:5000/clid10/noflag-cockroachdb-cockroach-helm-operator
    source: quay.io/cockroachdb/cockroach-helm-operator
  - mirrors:
    - localhost:5000/clid10/noflag-helmoperators-cockroachdb
    source: quay.io/helmoperators/cockroachdb
status: {}
```
